### DISCUSSION TITLE
Why does this work for .jpg but not for .png?

### DISCUSSION THREAD

**Discussion by goemonburo** [ImageMagick/ImageMagick#7390] [General]
I made a little script to add a white border to images but can't figure out why it only works for .jpgs.  Anyone know what I'm doing wrong? 

```
#! /bin/bash
for FILE in "$@";	    
do mogrify -bordercolor white -border 400x400 -path /home/myhome/borderadded $FILE;
done;
```

This works perfectly for any jpg...I find it, right click, choose "Add Border" from the pulldown menu...and it plops it into the /borderadded folder with a 400x400 white border.  

But it doesn't work for pngs.  

I'm running Lubuntu 22.04LTS.

**Comment by fmw42** [comment]
Do mogrify on a whole folder rather than file by file.

```
#! /bin/bash
cd folder_containing_images	 
mkdir /home/myhome/borderadded # if it does not already exist
mogrify -path /home/myhome/borderadded -bordercolor white -border 400x400  *;

```
But you don't need a script to do this.  You can just do the commands in a terminal.

**Comment by goemonburo** [comment]
Thank you, yes, I could.  Except that I would prefer to choose individual files since some I want borders and some I don't.  Does anyone know why the above (original) script works for jpgs but not for pngs?

**Reply by fmw42** [reply]
Where are you putting the list of files that you want to use? $@ would be every file or at least every file in your working directory. So how does that select the files you want?

**Comment by goemonburo** [comment]
The file is designed to use the GUI, so I cntl-click on 1, 2, or however many images I want to add borders to, then right click, and a "AddBorder" option is in the pulldown.  It operates the above script on any files that I've selected.  But weirdly only works for jpgs.

**Reply by fmw42** [reply]
Out of curiosity what GUI?

**Comment by snibgo** [comment]
> mogrify -bordercolor white -border 400x400 -path /home/myhome/borderadded $FILE

As a general rule: if you have only one input image, don't use `magick mogrify` (or just `mogrify`). Instead, use just `magick`.

`magick mogrify` is used when you have multiple inputs, and you want to apply the same processing to each of them.

**Comment by goemonburo** [comment]
@snibgo Thank you.  Generally the program is working on multiple images.  And again, I'm really just asking the "why doesn't it work on .png" files.  Does anyone here know?  

As far as the GUI, it's whatever the Lubuntu default is.  PCManFM-Qt, I think?

**Comment by fmw42** [comment]
I don't have your GUI, so I put some JPGs and PNGs in a folder test1 and ran

```
cd
cd desktop/test1
for FILE in "*";	    
do mogrify -bordercolor white -border 400x400 -path ../test2 $FILE
done

```

That worked perfectly. So Imagemagick is working properly.

So I would suggest you check what is in $@. Perhaps you GUI is not capturing PNGs for some reason.

What do you get from 

```
for FILE in "$@";	    
do echo $file
done
```

Does it list all the images that you think you have selected?

**Comment by snibgo** [comment]
>  And again, I'm really just asking the "why doesn't it work on .png" files. Does anyone here know?

It should work. Please show a PNG file for which this doesn't work.

**Comment by goemonburo** [comment]
Thank you (everyone!) for sticking with me on this.  Yes, I know, it really SHOULD work, and I'm pretty sure that it is indeed something in the GUI that's causing the issue.  Does anyone know how to print the name of a file in the above to a document, rather than using "echo" to a screen?  Because this is a program called by right clicking and choosing the option from a pull down, it's not being run in a terminal.  

I thought "cat $file >> textfile.txt" would work, but it is creating the file but writing nothing, neither for pngs nor for jpgs.  I suspect when I have the syntax right I'll see that the only files being selected are indeed the jpgs, as was surmised above by @fmw42 .

Thank you again!

**Comment by snibgo** [comment]
I suggest something like:
```
echo $file >>mylog.txt
```

**Reply by goemonburo** [reply]
That's what I had, and it's not working.  That was exactly what I put.  Tried with and without a space after the >> too....
(It doesn't work regardless of whether the file is mogrified or not.)

**Comment by goemonburo** [comment]
Okay, so...my wording wasn't clear:  The above "echo $file >>mylog.txt" puts the PHOTO into the file.  

What I want is it to put the filename string.  When I open mylog.txt it's a garbled mess of ascii...because the photo itself is being echoed...

**Reply by fmw42** [reply]
Your GUI is selecting the photo not the name. Look into your GUI to see what it has for controls and what the selector does.  Sorry, this is not an Imagemagick problem.  You need experts on `$@` and how that receives the information from your GUI.  Perhaps some Linux expert can comment.  Does it work differently if you use `$*` rather than `$@`?  I am not a bash unix expert.

**Comment by snibgo** [comment]
> The above "echo $file >>mylog.txt" puts the PHOTO into the file.

As fmw42 says. Your `$file` is being set to the contents of the file (or files), instead of the filename (or filenames).

IM can use the contents of a single file, like this:
```
cat $file | magick - -bordercolor white -border 400x400 out.jpg
```

**Comment by goemonburo** [comment]
Got it:  printf not cat does the trick!  (Now to see what's happening...)

**Comment by goemonburo** [comment]
Figured it out.  Filenames with spaces mess everything up.  Can't believe it took so long for me to figure this out!  Thank you for the help!

**Reply by fmw42** [reply]
Yes, filename with spaces need quotes around the complete names.

**Comment by goemonburo** [comment]
Will changing $FILE to "$FILE" fix it or am I needing to do it some other way?

(I think I'll just make sure there aren't spaces, that's easy to script.)

**Reply by fmw42** [reply]
That might work.  But I am concerned that since I do not understand how your are converting files to filenames that any space in your $@ might be considered a separate file. So putting quotes about the $FILE might be too late.  Give it a try and see what happens. If putting quotes about $FILE is insufficient, then I do not know what to suggest. But it is always advisable to put the quotes about the use of $FILE.

**Comment by goemonburo** [comment]
Thank you, All!  Rather than tie up your time trying to bash script when this is an Imagemagick discussion, I've just opted to do a separate script that renames files in a folder, stripping out the spaces and swapping them with underscores.  It isn't elegant but it works just fine.  I think I'm all set now!  Thank you so much for the help.

**Reply by fmw42** [reply]
Did you even try simply your script where you modified it to have "$FILE" with quotes?  Did that work?

**Comment by goemonburo** [comment]
Hmmm.  It appears the quotes did the trick.  You sounded doubtful that it would, so I thought I'd spare you all the trouble of messing around more and just rename the files...but it works just fine!  Very solid fix!  Thank you!

**Reply by fmw42** [reply]
Interesting and educational.  Thanks

**Comment by goemonburo** [comment]
Sorry to tack on one more question that may be also a bash shell one but I've tried everything and am wondering if mogrify has different syntax:  

Currently the script as you can see saves all images to -path ../test2.  I'd like to have it rename the file to a new name such as "with-border$FILE" and keep it in the same directory.  Is there any way to do that?  It seems none of the bash-related ways to do this are working...Does mogrify always only overwrite the original file?

**Reply by fmw42** [reply]
As far as I know that is correct. It overwrites unless you use -path to a separate directory. But an IM developer might comment or another user who know how to do what you want.

**Reply by goemonburo** [reply]
Thank you.  No worries then.  I can easily deal with just using the path.  Since this is a shell script I suppose I could just use a mv command to move the file immediately from that folder to the current one, and append or prepend what I want to the name so it doesn't overwrite.
